keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,540 @@
+      - [Beta -&gt; GA Graduation](#beta---ga)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Implementation History](#implementation-history)
+- [Drawbacks [optional]](#drawbacks-optional)
+- [Alternatives [optional]](#alternatives-optional)
+<!-- /toc -->
+
+## Summary
+
+Kubernetes natively supports dynamic provisioning for file and block storage but lacks support for object bucket provisioning.
+Without a formal API to describe and manage bucket provisioning, storage providers are left to write their own provisioners including all of the necessary controller code.
+This "wild west" landscape for bucket provisioning is potentially confusing, overly complex, inconsistent, may slow down Kubernetes adoption by object store users, and may hinder application portability.
+
+We propose a common Kubernetes control-plane API for the management of object store bucket lifecycles which has no dependencies on the underlying providers.
+
+## Motivation
+
+By defining a Kubernetes API for object bucket provsioning, we provide a consistent framework for all object store providers and thus enhance the end user's storage experience. This API allows object store vendors to focus on how they wish to provision buckets instead of writing and testing the controller aspects of provisioning.
+
+### Goals
+
++ Define a _control-plane_ object bucket management API thus relieving object store providers from Kubernetes controller details.
++ Minimize technical ramp-up for storage vendors who have already contributed external storage provisioners (file and block).
++ Make bucket consumption similar to file or block consumption using familiar concepts and commands.
++ Use native Kubernetes resources where possible, again, to keep the bucket experience for users and admins similar to existing storage provisioning.
++ Be unopinionated about the underlying object-store and at the same time provide a flexible API such that provisioner specific features can be supported.
++ Ensure bucket consuming pods wait until the target buckets have been created and are accessible. Thus there is no specific order required for when a bucket claim is created vs. when the app pod is run.
++ Present similar user and admin experiences for both _greenfield_ (new) and _brownfield_ (existing) bucket provisioning.
+
+### Non-Goals
+
++ Update the native Kubernetes PVC-PV API to support object buckets.
++ Define a native _data-plane_ object store API.
++ Handle the small percentage of apps that will not be portable due to use of non-compatible object-store features.
+
+## Proposal
+
+Create a bucket provisioning library, similar to the Kubernetes [sig-storage-lib-external-provisioner](https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/blob/master/controller/controller.go) library, which handles all of the controller and Kubernetes resources related to provisioning.
```

Comment:
After years of dealing with the issues of having multiple vendors support provisioning of volumes, the community decided on an approach that abstracts the control plane behind a simple gRPC API, allowing for provisioning drivers to be effectively independent of Kubernetes.  I believe that this is the approach we should also use for object.  Instead of providing a library and asking object vendors to implement a controller using the library, I think we should provide top-level and sidecar controllers and define the interface via gRPC.  This massively simplifies the process of writing and deploying a driver.  Using a central top-level controller (e.g., see the 1.17 volume snapshot architecture) allows for a clean approach to CRD versioning and lifecycle management and standardizing observability metrics.

It also could theoretically allow for situations in the future where the gRPC functions could be optionally added to both CSI (for cases where volume storage providers *also* provide object support) and separately (e.g., via a new "OSI").

Reply:
**I'd like to see discussion on this suggestion.**
I am not experienced enough in CSI to offer arguments to this comment. There are a couple of provsioners using this lib and my understanding is that the effort was relatively low.

Reply:
Having someone deliver the whole solution in a single controller creates a lot of problems.  For example:

  * If there is a bug in the library that is discovered, how do you ensure that all the controllers using that library are updated?

  * If you have multiple controllers installed, who "owns" the CRDs?  How do they get installed/upgraded?

  * If storage vendors are providing their own whole controllers, how do you ensure a degree of uniformity (and correctness) around observability?

The volume snapshot folks directly addressed this set of issues with the beta volume snapshot work and came up with a dual-control + driver architecture:

  * A single-instance global controller that handles most of the complicated creation, provisioner lookup, and binding logic

  * A very simple sidecar controller that serves primarily to convert k8s resources into gRPC calls

  * A plugin driver supplied by storage vendors that is kubernetes and language independent.

I'm advocating a very similar model here.

Reply:
+1 And since many storage vendors have already switched over to CSI drivers, it's going to be a hard sell to convince them they should adopt a model that k8s has moved on from.

Reply:
I am starting to value a CSI-like approach (maybe COSI or "cozy?).
@alarge can you elaborate on this point:
> If there is a bug in the library that is discovered, how do you ensure that all the controllers using that library are updated?

I agree it's a potential problem with the library. But CRDs can support multiple versions and so can the lib (thought it doesn't today). Can you explain how CSI mitigates this issue?

Reply:
@jeffvance A bug in this context could mean 2 kinds of bug: 1) kubernetes related bug(i.e CRD, controller common logic in the library) 2) object storage vendor related bug(i.e Provisioner bug). The library approach IMO seems to be unnecessarily coupling these 2 types of bugs. An issue on either side(K8s controller/CRD VS. Object Storage Provisioner) IMO shouldn't require attention from the other side. For example, a common library bug fix shouldn't require vendor to release a new version of their controller because there's no change in the "Provisioner" they've implemented. Having a CSI-like interface would allow us to decouple these 2 sides of the problem. For instance, a bug fix in the central controller would just require end-user to bump up the version number in the deployment yaml and object storage vendor is no involved at all.

Reply:
I do like the separation of provisioner bugs vs a csi bug. I'll add this point to the `CSI Pros` section of the KEP.

Reply:
Follow up: it [appears](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/csi/csi_client.go#L199) that `accessMode` is used by CSI. I don't yet know how it's enforced or controlled, but it is passed along in the request.